README.md
# Polling Request [![Build Status](https://secure.travis-ci.org/jch/polling_request.png)](http://travis-ci.org/jch/polling_request)

jQuery helper object for AJAX polling a [progress-aware endpoint](#progress-aware-endpoint).

## Usage

```coffeescript
req = new PollingRequest
  url: "/states.csv"
  interval: 1000   # polling interval in milliseconds
  progress: (n)->
    console.log "Progress: #{n} percent"
  success: (res)->
    console.log "Response body: #{res}"
  error: (status, response)->
    console.log("Status: #{status}, Body: #{response}")
  complete: (status, response)->
    console.log("This is called after everything else")

req.status   # 'pending'
req.start()
req.status   # 'running'
req.progress # 0

# sometime later
req.status   # 'running'
req.progress # 50

# much later
req.status   # 'success'
req.progress # 100

# example of setting a hard timeout
setTimeout ->
  req.stop()
, 5000
```

## Progress Aware Endpoint

PollingRequests will have a header `X-POLLING-REQUEST` set. It expects an HTTP
endpoint to respond with a [202 Accepted][202] status code when there is more
processing to be done. The JSON response body can optionally include a
`progress` property whose value is an integer 0..100 inclusive to indicate the
percentage of processing completed.

```bash
$ curl -i http://localhost/endpoint
HTTP/1.1 202 Accepted
{ progress: 0 }

$ curl -i http://localhost/endpoint
HTTP/1.1 202 Accepted
{ progress: 50 }

$ curl -i http://localhost/endpoint
HTTP/1.1 200 Success
Content-Type: text/plain
Successful responses can be any content type
```

[202]: http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.2.3

## Rails

To access PollingRequest from the asset pipeline, add the following to your
Gemfile:

```ruby
gem 'polling_request'
```

And this to your asset manifest:

```javascript
//= require polling_request
```

## Bower

PollingRequest is in the [Bower registry](http://sindresorhus.com/bower-components/).  JavaScript and CoffeeScript sources are included.

```bash
$ bower install polling-request
```

## Development

Dependencies are managed by [npm](https://npmjs.org/) and [Bower](http://bower.io/).

```bash
$ npm install
```

```bash
$ bower install
```

Tests are written with QUnit and can be run in a browser by opening
`test/test.html`. Continuous integration is run with phantomjs:

```bash
$ ./node_modules/phantomjs/lib/phantom/bin/phantomjs test/runner.js test/test.html
```

## Wish List

* Better docs for building and installing project

## Contributing

1. [Fork it](https://help.github.com/articles/fork-a-repo)
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](https://help.github.com/articles/using-pull-requests)


CHANGELOG.md
# CHANGELOG

## 0.0.3

* Bower compatibility
* Travis CI

## 0.0.2

* added `stopped` callback

## 0.0.1

* initial public release


test/runner.js
/*
 * QtWebKit-powered headless test runner using PhantomJS
 *
 * PhantomJS binaries: http://phantomjs.org/download.html
 * Requires PhantomJS 1.6+ (1.7+ recommended)
 *
 * Run with:
 *   phantomjs runner.js [url-of-your-qunit-testsuite]
 *
 * e.g.
 *   phantomjs runner.js http://localhost/qunit/test/index.html
 */

/*global phantom:false, require:false, console:false, window:false, QUnit:false */

(function() {
  'use strict';

  var url, page, timeout,
    args = require('system').args;

  // arg[0]: scriptName, args[1...]: arguments
  if (args.length < 2 || args.length > 3) {
    console.error('Usage:\n  phantomjs runner.js [url-of-your-qunit-testsuite] [timeout-in-seconds]');
    phantom.exit(1);
  }

  url = args[1];
  page = require('webpage').create();
  if (args[2] !== undefined) {
    timeout = parseInt(args[2], 10);
  }

  // Route `console.log()` calls from within the Page context to the main Phantom context (i.e. current `this`)
  page.onConsoleMessage = function(msg) {
    console.log(msg);
  };

  page.onInitialized = function() {
    page.evaluate(addLogging);
  };

  page.onCallback = function(message) {
    var result,
      failed;

    if (message) {
      if (message.name === 'QUnit.done') {
        result = message.data;
        failed = !result || result.failed;

        phantom.exit(failed ? 1 : 0);
      }
    }
  };

  page.open(url, function(status) {
    if (status !== 'success') {
      console.error('Unable to access network: ' + status);
      phantom.exit(1);
    } else {
      // Cannot do this verification with the 'DOMContentLoaded' handler because it
      // will be too late to attach it if a page does not have any script tags.
      var qunitMissing = page.evaluate(function() { return (typeof QUnit === 'undefined' || !QUnit); });
      if (qunitMissing) {
        console.error('The `QUnit` object is not present on this page.');
        phantom.exit(1);
      }

      // Set a timeout on the test running, otherwise tests with async problems will hang forever
      if (typeof timeout === 'number') {
        setTimeout(function() {
          console.error('The specified timeout of ' + timeout + ' seconds has expired. Aborting...');
          phantom.exit(1);
        }, timeout * 1000);
      }

      // Do nothing... the callback mechanism will handle everything!
    }
  });

  function addLogging() {
    window.document.addEventListener('DOMContentLoaded', function() {
      var currentTestAssertions = [];

      QUnit.log(function(details) {
        var response;

        // Ignore passing assertions
        if (details.result) {
          return;
        }

        response = details.message || '';

        if (typeof details.expected !== 'undefined') {
          if (response) {
            response += ', ';
          }

          response += 'expected: ' + details.expected + ', but was: ' + details.actual;
        }

        if (details.source) {
          response += "\n" + details.source;
        }

        currentTestAssertions.push('Failed assertion: ' + response);
      });

      QUnit.testDone(function(result) {
        var i,
          len,
          name = result.module + ': ' + result.name;

        if (result.failed) {
          console.log('Test failed: ' + name);

          for (i = 0, len = currentTestAssertions.length; i < len; i++) {
            console.log('    ' + currentTestAssertions[i]);
          }
        }

        currentTestAssertions.length = 0;
      });

      QUnit.done(function(result) {
        console.log('Took ' + result.runtime +  'ms to run ' + result.total + ' tests. ' + result.passed + ' passed, ' + result.failed + ' failed.');

        if (typeof window.callPhantom === 'function') {
          window.callPhantom({
            'name': 'QUnit.done',
            'data': result
          });
        }
      });
    }, false);
  }
})();

src/polling_request.js
// Generated by CoffeeScript 1.6.3
(function() {
  var PollingRequest;

  PollingRequest = (function() {
    function PollingRequest(options) {
      var _base, _base1, _base2, _base3, _base4;
      this.options = options;
      this.progress = 0;
      this.status = 'pending';
      (_base = this.options).interval || (_base.interval = 1000);
      (_base1 = this.options).success || (_base1.success = function() {});
      (_base2 = this.options).progress || (_base2.progress = function() {});
      (_base3 = this.options).error || (_base3.error = function() {});
      (_base4 = this.options).stopped || (_base4.stopped = function() {});
    }

    PollingRequest.prototype.start = function() {
      var _this = this;
      if (this.timer) {
        return;
      }
      return this.timer = setInterval(function() {
        return _this._run();
      }, this.options.interval);
    };

    PollingRequest.prototype.stop = function() {
      this.status = 'stopped';
      clearInterval(this.timer);
      this.timer = null;
      return this.options.stopped();
    };

    PollingRequest.prototype.status = function() {
      return this.status;
    };

    PollingRequest.prototype.progress = function() {
      return this.progress;
    };

    PollingRequest.prototype._run = function() {
      var _this = this;
      return $.ajax({
        url: this.options.url,
        cached: false,
        accepts: {
          json: 'application/json'
        },
        beforeSend: function(jqXHR, settings) {
          return _this.status = 'running';
        },
        statusCode: {
          200: function(data, textStatus, jqXHR) {
            _this.progress = 100;
            _this.options.success(data);
            return _this.stop();
          },
          202: function(data, textStatus, jqXHR) {
            _this.progress = data.progress || 0;
            return _this.options.progress(_this.progress);
          }
        },
        error: function(jqXHR, textStatus, errorThrown) {
          _this.options.error(jqXHR.status, jqXHR.responseText);
          return _this.stop();
        }
      });
    };

    return PollingRequest;

  })();

  (typeof exports !== "undefined" && exports !== null ? exports : this).PollingRequest = PollingRequest;

}).call(this);
